File: python/sparktk/graph/ops/page_rank.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#


def page_rank(self, convergence_tolerance=None, reset_probability=None, max_iterations=None):
    """
    **Page Rank**

    Page Rank is a popular statistic that ranks vertices based off of
    connectivity in the global graph

    Exactly 1 of convergence_tolerance and max_iterations must be set (termination criteria)

    Parameters
    ----------

    :convergence_tolerance: (Float) If the difference between successive iterations is less than this, the algorithm terminates. Mutually exclusive with max_iterations
    :reset_probability: (Float) Value for the reset probabiity in the page rank algorithm
    :max_iterations: (Int) Maximum number of iterations the page rank should run before terminating. Mutually exclusive with convergence_tolerance

    :return: (Frame) Frame containing the vertex id's and their page rank 

    Examples
    --------

        >>> vertex_schema = [('id', int)]
        >>> edge_schema = [('src', int), ('dst', int)]

        >>> vertex_rows = [ [1], [2], [3], [4], [5] ]
        >>> edge_rows = [ [1, 2], [1, 3], [2, 3], [1, 4], [4, 5] ]
        >>> vertex_frame = tc.frame.create(vertex_rows, vertex_schema)
        >>> edge_frame = tc.frame.create(edge_rows, edge_schema)

        >>> graph = tc.graph.create(vertex_frame, edge_frame)

        >>> result = graph.page_rank(max_iterations=20)
        >>> result.inspect()
        [#]  id  pagerank
        =================
        [0]   1      0.15
        [1]   2    0.1925
        [2]   3  0.356125
        [3]   4    0.1925
        [4]   5  0.313625

    """
    from sparktk.frame.frame import Frame
    return Frame(self._tc, self._scala.pageRank(
        self._tc.jutils.convert.to_scala_option(max_iterations),
        self._tc.jutils.convert.to_scala_option(reset_probability),
        self._tc.jutils.convert.to_scala_option(convergence_tolerance)))

File: sparktk-core/src/main/scala/org/trustedanalytics/sparktk/graph/internal/ops/PageRank.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.graph.internal.ops

import org.trustedanalytics.sparktk.frame.Frame
import org.apache.spark.sql.functions.{ sum, array, col, count, explode, struct }
import org.graphframes.GraphFrame
import org.apache.spark.sql.DataFrame
import org.graphframes.lib.AggregateMessages

import org.apache.spark.sql.Column
import org.apache.spark.sql.functions.lit

import org.trustedanalytics.sparktk.graph.internal.{ GraphState, GraphSummarization, BaseGraph }

trait PageRankSummarization extends BaseGraph {
  /**
   * Calculates the page rank for each vertex
   *
   *    **Basics and Background**
   *
   *    *PageRank* is a method for determining which vertices in a directed graph are
   *    the most central or important.
   *    *PageRank* gives each vertex a score which can be interpreted as the
   *    probability that a person randomly walking along the edges of the graph will
   *    visit that vertex.
   *
   *    The calculation of *PageRank* is based on the supposition that if a vertex has
   *    many vertices pointing to it, then it is "important",
   *    and that a vertex grows in importance as more important vertices point to it.
   *    The calculation is based only on the network structure of the graph and makes
   *    no use of any side data, properties, user-provided scores or similar
   *    non-topological information.
   *
   *    *PageRank* was most famously used as the core of the Google search engine for
   *    many years, but as a general measure of :term:`centrality` in a graph, it has
   *    other uses to other problems, such as :term:`recommendation systems` and
   *    analyzing predator-prey food webs to predict extinctions.
   *
   *    **Background references**
   *
   *   Basic description and principles: `Wikipedia\: PageRank`_
   *   Applications to food web analysis: `Stanford\: Applications of PageRank`_
   *   Applications to recommendation systems: `PLoS\: Computational Biology`_
   *
   * NOTE:: exactly one of maxIterations and convergenceTolerance must be set
   *
   * @param maxIterations The maximum number of iterations to run for. Mutually exclusive with convergenceTolerance
   * @param resetProbability The reset probability for the page rank equation
   * @param convergenceTolerance If the difference between two iterations of page rank is less than this, page rank will terminate. Mutually exclusive with maxIterations
   * @return The dataframe containing the vertices and their corresponding page rank
   */
  def pageRank(maxIterations: Option[Int], resetProbability: Option[Double], convergenceTolerance: Option[Double]): Frame = {
    execute[Frame](PageRank(maxIterations, resetProbability, convergenceTolerance))
  }
}

case class PageRank(maxIterations: Option[Int], resetProbability: Option[Double], convergenceTolerance: Option[Double]) extends GraphSummarization[Frame] {

  override def work(state: GraphState): Frame = {
    require(maxIterations.isEmpty || convergenceTolerance.isEmpty, "Only set one of max iterations or convergence tolerance")
    require(!maxIterations.isEmpty || !convergenceTolerance.isEmpty, "Must set one of max iterations or convergence tolerance")

    val pageRank = state.graphFrame.pageRank

    // Set the relevant values

    val setTol = convergenceTolerance.fold(pageRank)(pageRank.tol)

    val setMaxIterations = maxIterations.fold(pageRank)(pageRank.maxIter)

    val setReset = resetProbability.fold(pageRank)(pageRank.resetProbability)

    new Frame(pageRank.run().vertices)
  }
}
